### ISSUE TITLE
Batch index offset with -sample in train

### ISSUE THREAD

**Issue by francoishernandez** [OpenNMT/OpenNMT#227]
Hi,
I am getting a strange behavior while using -sample option in training. The number of batches computed is 15649 whereas 15652 batches were prepared. In that configuration this might run properly (at least until batch 15649), but previously I had the opposite configuration with only 15649 batches prepared for a predicted number of 15650 and it crashed (batch idx out of range).
Where could this come from ?

[04/24/17 18:51:16 INFO] Using GPU(s): 2
[04/24/17 18:51:16 WARNING] The caching CUDA memory allocator is enabled. This allocator improves performance at the cost of a higher GPU memory usage. To optimize for memory, consider disabling it by setting the environment variable: THC_CACHING_ALLOCATOR=0
[04/24/17 18:51:16 INFO] Training Sequence to Sequence with Attention model...
[04/24/17 18:51:16 INFO] Loading data from '../speech2textONMT/phone2text/preprocess-train.t7'...
[04/24/17 18:54:55 INFO]  * sampling 1000000 instances from 6540582 at each epoch
[04/24/17 18:55:27 INFO]  * vocabulary size: source = 44; target = 50003
[04/24/17 18:55:27 INFO]  * additional features: source = 0; target = 0
[04/24/17 18:55:27 INFO]  * maximum sequence length: source = 50; target = 33
[04/24/17 18:55:27 INFO]  * number of training sentences: 6540582
[04/24/17 18:55:27 INFO]  * number of batches: 15649
[04/24/17 18:55:27 INFO]    - source sequence lengths: equal
[04/24/17 18:55:27 INFO]    - maximum size: 64
[04/24/17 18:55:27 INFO]    - average size: 417.96
[04/24/17 18:55:27 INFO]    - capacity: 100.00%
[04/24/17 18:55:27 INFO] Building model...
[04/24/17 18:55:27 INFO]  * Encoder:
[04/24/17 18:55:27 INFO]    - word embeddings size: 128
[04/24/17 18:55:27 INFO]    - type: pyramidal deep bidirectional
[04/24/17 18:55:27 INFO]    - structure: cell = LSTM; layers = 3; rnn_size = 600; dropout = 0.3
[04/24/17 18:55:27 INFO]  * Decoder:
[04/24/17 18:55:28 INFO]    - word embeddings size: 256
[04/24/17 18:55:28 INFO]    - attention: global (general)
[04/24/17 18:55:28 INFO]    - structure: cell = LSTM; layers = 3; rnn_size = 600; dropout = 0.3
[04/24/17 18:55:30 INFO] Current sampling does not require individual loss
[04/24/17 18:55:30 INFO] Initializing parameters...
[04/24/17 18:55:32 INFO]  * number of parameters: 68250203
[04/24/17 18:55:32 INFO] Preparing memory optimization...
[04/24/17 18:55:33 INFO]  * sharing 70% of output/gradInput tensors memory between clones
[04/24/17 18:55:33 INFO] Start training...
[04/24/17 18:55:33 INFO] 
[04/24/17 18:55:33 INFO] Sampling...
[04/24/17 18:55:41 INFO] Created sampler...
[04/24/17 18:55:43 INFO] Sampled 1000000 instances
[04/24/17 18:55:47 INFO] Prepared 15652 batches
[04/24/17 18:59:17 INFO] Epoch 1 ; Iteration 1000/15652 ; Learning rate 1.0000 ; Source tokens/s 8994 ; Perplexity 15319.05

**Comment by francoishernandez** [comment]
Update:
The following error happened while running the training which initiation is presented above.
Any ideas?

bad argument #2 to '?' (out of range at /home/moses/torch/pkg/torch/generic/Tensor.c:913)
stack traceback:
	[C]: at 0x7f6a74eb3dc0
	[C]: in function '__index'
	./onmt/train/Trainer.lua:138: in function 'getBatchIdx'
	./onmt/train/Trainer.lua:170: in function 'trainEpoch'
	./onmt/train/Trainer.lua:351: in function 'train'
	train.lua:173: in function 'main'
	train.lua:178: in main chunk
	[C]: in function 'dofile'
	...oses/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:150: in main chunk
	[C]: at 0x00406670

**Comment by vince62s** [comment]
@guillaumekln I think your commit https://github.com/OpenNMT/OpenNMT/commit/7ff9a29808dcf59c4e1a7ffac845cdda2ac4ce76#diff-994c78c41703e95cf778bb21b272cc4b
introduced that bug.
I could run without any issue with a hash prior to this commit.
I think in some instance the number of batch calculated before by batchcount() and after sampling by data:sample() differ and the dimension of the batchOrder tensor is not accurate (misses 1)

**Comment by guillaumekln** [comment]
Thanks, I will fix that.

In the meantime you could just use an older commit: 6521cc0.
